C# and Go visitors emit non-compiling class renderings (self-assignment on shadowed fields; constructor naming mismatch; pointer mismatches in assignments)

### Context

While investigating aws/aws-cdk#38452 (the `aws-s3`/`aws-kms` README example implementing `IResourceWithPolicyV2` as an object literal, which translates into broken code in every non-TS language), rewriting the example as a named class — the fix confirmed there, see aws/aws-cdk#38461 — exposed renderer bugs in the C# and Go visitors that apply to *any* translated class, not just that README. Filing here as the rosetta-side follow-up suggested in that issue's triage.

Minimal repro (a snippet-local class whose constructor parameter shares a name with a private field):

```ts
class MyResource {
  public readonly env: string;
  private readonly resource: string;

  constructor(resource: string) {
    this.resource = resource;
    this.env = 'production';
  }

  public description(): string {
    return this.resource;
  }
}

const r = new MyResource('bucket');
```

### C#: dropped `this.` produces a self-assignment

The C# visitor suppresses `this.` unconditionally. Public members are PascalCased so they can never collide with a camelCase parameter, but **private fields keep their camelCase name** — and then the constructor assignment degrades into a self-assignment (compiler warning CS1717; the field is never set).

**Before:**

```csharp
public MyResource(string resource)
{
    resource = resource;        // CS1717 — field never assigned
    Env = "production";
}
```

**New** (keep `this.` only when the member's rendered name is shadowed by a parameter of an enclosing function; all other output unchanged):

```csharp
public MyResource(string resource)
{
    this.resource = resource;
    Env = "production";
}
```

### Go: constructor call/definition naming mismatch

`constructorDeclaration` renders `new`/`New` according to the class's exportedness, but `newExpression` (the call site) hardcodes `New` — the code even carries a comment asking `// Should this be "new" if the class is unexported?`. Result: the definition says `func newMyResource` while the call site invokes the non-existent `NewMyResource`.

**Before:**

```go
r := NewMyResource(jsii.String("bucket"))   // NewMyResource does not exist
```

**New:**

```go
r := newMyResource(jsii.String("bucket"))
```

Caveat handled in the fix: corpus/test snippets declare unexported `/// fake-from-jsii` classes that stand in for library types — those must keep the exported `NewVpc(...)` call form, since only the call site renders. The fix distinguishes genuinely snippet-local classes (unexported *and* not jsii-resolvable) from classes modeling library types.

### Go: assignments put the wrong side of the pointer on the right

Parameters and properties are both pointer-valued in Go, but they do not *render* alike:

- a **property** renders as the pointer field itself — `this.env`, a `*string`;
- a **parameter** renders dereferenced — `*name`, a `string`.

So the two need opposite right-hand sides: a property target wants a pointer (leave a pointer alone, wrap a literal), a parameter target wants a value (dereference a pointer, leave a literal raw). The visitor treats them as one case — it suppresses dereferencing on the RHS for both (`isPtrAssignmentRValue`) and wraps literals for neither — so each is broken in one direction:

| TypeScript | Before | Why it fails |
| --- | --- | --- |
| `this.env = 'production'` | `this.env = "production"` | `*string = string` |
| `next = other` (both parameters) | `*next = other` | `string = *string` |

**New** — both flags key off whether the assignment *target* renders as a pointer, which only a property does:

| TypeScript | New |
| --- | --- |
| `this.env = 'production'` | `this.env = jsii.String("production")` |
| `this.env = env` | `this.env = env` |
| `name = 'fallback'` | `*name = "fallback"` |
| `next = other` | `*next = *other` |

Python and Java render all of the above correctly today.

### Fixes

All three, developed test-first with translations-corpus snippets pinning all four languages (so the diff proves no change alters another target's output), are up as PR #3773. Visitor `VERSION`s are bumped to invalidate cached translations.

--- a/src/languages/csharp.ts
+++ b/src/languages/csharp.ts
@@ -77,7 +77,7 @@ export class CSharpVisitor extends DefaultVisitor<CSharpLanguageContext> {
    * Bump this when you change something in the implementation to invalidate
    * existing cached translations.
    */
-  public static readonly VERSION = '1';
+  public static readonly VERSION = '2';
 
   public readonly language = TargetLanguage.CSHARP;
 
@@ -340,12 +340,35 @@ export class CSharpVisitor extends DefaultVisitor<CSharpLanguageContext> {
 
     // Suppress the LHS of the dot operator if it's "this." (not necessary in C#)
     // or if it's an imported module reference (C# has namespace-wide imports).
+    // Exception: keep "this." when the member, as it will be rendered, is
+    // shadowed by a parameter of an enclosing function. This matters for
+    // private fields (which keep their camelCase name): in a constructor
+    // assigning a same-named field, dropping the qualifier degrades
+    // `this.resource = resource` into the self-assignment `resource = resource;`.
+    const dropThis = lhs === 'this' && !this.memberShadowedByParameter(node, renderer);
     const objectExpression =
-      lhs === 'this' || this.dropPropertyAccesses.has(lhs)
+      dropThis || this.dropPropertyAccesses.has(lhs)
         ? []
         : [renderer.updateContext({ propertyOrMethod: false }).convert(node.expression), '.'];
 
     return new OTree([...objectExpression, renderer.updateContext({ propertyOrMethod: true }).convert(node.name)]);
+  }
+
+  private memberShadowedByParameter(node: ts.PropertyAccessExpression, renderer: CSharpRenderer): boolean {
+    const text = node.name.text;
+    // The name the member will render as: private properties keep their
+    // camelCase name, everything else is PascalCased (see `identifier`).
+    const renderedName = renderer.currentContext.privatePropertyNames.includes(text) ? text : ucFirst(text);
+
+    for (let current: ts.Node | undefined = node.parent; current != null; current = current.parent) {
+      if (
+        ts.isFunctionLike(current) &&
+        current.parameters.some((p) => ts.isIdentifier(p.name) && p.name.text === renderedName)
+      ) {
+        return true;
+      }
+    }
+    return false;
   }
 
   public override parameterDeclaration(node: ts.ParameterDeclaration, renderer: CSharpRenderer): OTree {

--- a/src/languages/go.ts
+++ b/src/languages/go.ts
@@ -631,20 +631,23 @@ export class GoVisitor extends DefaultVisitor<GoLanguageContext> {
   public override binaryExpression(node: ts.BinaryExpression, renderer: AstRenderer<GoLanguageContext>): OTree {
     if (node.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
       const symbol = symbolFor(renderer.typeChecker, node.left);
-      // Parameters and properties are pointer-valued: a pointer-valued RHS
-      // must not be dereferenced (isPtrAssignmentRValue), and a literal RHS
-      // must be wrapped (wrapPtr) — `this.env = "production"` needs
-      // `jsii.String("production")`, not a raw literal.
-      const pointerTarget =
-        symbol?.valueDeclaration != null &&
-        (ts.isParameter(symbol.valueDeclaration) || ts.isPropertyDeclaration(symbol.valueDeclaration));
+      // What the right-hand side must be follows from how the left-hand side
+      // *renders*, and that differs between the two pointer-valued
+      // declarations. A property renders as the pointer field itself
+      // (`this.env`), so the RHS has to be a pointer: leave a pointer
+      // undereferenced, and wrap a literal — `this.env = "production"` needs
+      // `jsii.String("production")`. A parameter renders dereferenced (`*name`,
+      // see goName at the bottom of this file), so the target is a value and
+      // the RHS has to be one too — exactly the opposite of a property.
+      const target = symbol?.valueDeclaration;
+      const targetRendersAsPointer = target != null && ts.isPropertyDeclaration(target);
       return new OTree([
         renderer.convert(node.left),
         ' = ',
         renderer
           .updateContext({
-            isPtrAssignmentRValue: pointerTarget,
-            wrapPtr: pointerTarget || renderer.currentContext.wrapPtr,
+            isPtrAssignmentRValue: targetRendersAsPointer,
+            wrapPtr: targetRendersAsPointer,
           })
           .convert(node.right),
       ]);

--- a/src/languages/go.ts
+++ b/src/languages/go.ts
@@ -82,7 +82,7 @@ export class GoVisitor extends DefaultVisitor<GoLanguageContext> {
    * Bump this when you change something in the implementation to invalidate
    * existing cached translations.
    */
-  public static readonly VERSION = '1';
+  public static readonly VERSION = '2';
 
   public readonly indentChar = '\t';
 
@@ -191,11 +191,13 @@ export class GoVisitor extends DefaultVisitor<GoLanguageContext> {
   }
 
   public override newExpression(node: ts.NewExpression, renderer: GoRenderer): OTree {
-    const { classNamespace, className } = determineClassName.call(this, node.expression);
+    const { classNamespace, className, unexported } = determineClassName.call(this, node.expression);
     return new OTree(
       [
         ...(classNamespace ? [classNamespace, '.'] : []),
-        'New', // Should this be "new" if the class is unexported?
+        // Constructors of unexported (snippet-local) classes render unexported
+        // (see constructorDeclaration) — the call site must match.
+        unexported ? 'new' : 'New',
         className,
         '(',
       ],
@@ -203,7 +205,10 @@ export class GoVisitor extends DefaultVisitor<GoLanguageContext> {
       { canBreakLine: true, separator: ', ', suffix: ')' },
     );
 
-    function determineClassName(this: GoVisitor, expr: ts.Expression): { classNamespace?: OTree; className: string } {
+    function determineClassName(
+      this: GoVisitor,
+      expr: ts.Expression,
+    ): { classNamespace?: OTree; className: string; unexported?: boolean } {
       if (ts.isIdentifier(expr)) {
         // Imported names are referred to by the original (i.e: exported) name, qualified with the source module's go
         // package name.
@@ -225,7 +230,18 @@ export class GoVisitor extends DefaultVisitor<GoLanguageContext> {
           };
         }
 
-        return { className: ucFirst(expr.text) };
+        // A snippet-local class renders an unexported constructor. Classes
+        // that model library types (jsii-resolvable, incl. 'fake-from-jsii'
+        // test fixtures) keep the exported naming even when declared
+        // unexported in the snippet.
+        const classDeclaration = symbol?.declarations?.find(ts.isClassDeclaration);
+        return {
+          className: ucFirst(expr.text),
+          unexported:
+            classDeclaration != null &&
+            !isExported(classDeclaration) &&
+            lookupJsiiSymbolFromNode(renderer.typeChecker, expr) == null,
+        };
       }
       if (ts.isPropertyAccessExpression(expr)) {
         if (ts.isIdentifier(expr.expression)) {
@@ -615,14 +631,20 @@ export class GoVisitor extends DefaultVisitor<GoLanguageContext> {
   public override binaryExpression(node: ts.BinaryExpression, renderer: AstRenderer<GoLanguageContext>): OTree {
     if (node.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
       const symbol = symbolFor(renderer.typeChecker, node.left);
+      // Parameters and properties are pointer-valued: a pointer-valued RHS
+      // must not be dereferenced (isPtrAssignmentRValue), and a literal RHS
+      // must be wrapped (wrapPtr) — `this.env = "production"` needs
+      // `jsii.String("production")`, not a raw literal.
+      const pointerTarget =
+        symbol?.valueDeclaration != null &&
+        (ts.isParameter(symbol.valueDeclaration) || ts.isPropertyDeclaration(symbol.valueDeclaration));
       return new OTree([
         renderer.convert(node.left),
         ' = ',
         renderer
           .updateContext({
-            isPtrAssignmentRValue:
-              symbol?.valueDeclaration &&
-              (ts.isParameter(symbol.valueDeclaration) || ts.isPropertyDeclaration(symbol.valueDeclaration)),
+            isPtrAssignmentRValue: pointerTarget,
+            wrapPtr: pointerTarget || renderer.currentContext.wrapPtr,
           })
           .convert(node.right),
       ]);

--- a/test/translations/classes/constructor_parameter_shadows_field.cs
+++ b/test/translations/classes/constructor_parameter_shadows_field.cs
@@ -1,0 +1,18 @@
+class MyResource
+{
+    public string Env { get; }
+    private readonly string resource;
+
+    public MyResource(string resource)
+    {
+        this.resource = resource;
+        Env = "production";
+    }
+
+    public string Description()
+    {
+        return resource;
+    }
+}
+
+var r = new MyResource("bucket");

--- a/test/translations/classes/constructor_parameter_shadows_field.go
+++ b/test/translations/classes/constructor_parameter_shadows_field.go
@@ -1,0 +1,17 @@
+type myResource struct {
+	env *string
+	resource *string
+}
+
+func newMyResource(resource *string) *myResource {
+	this := &myResource{}
+	this.resource = resource
+	this.env = jsii.String("production")
+	return this
+}
+
+func (this *myResource) description() *string {
+	return this.resource
+}
+
+r := newMyResource(jsii.String("bucket"))

--- a/test/translations/classes/constructor_parameter_shadows_field.java
+++ b/test/translations/classes/constructor_parameter_shadows_field.java
@@ -1,0 +1,15 @@
+public class MyResource {
+    public final String env;
+    private final String resource;
+
+    public MyResource(String resource) {
+        this.resource = resource;
+        this.env = "production";
+    }
+
+    public String description() {
+        return this.resource;
+    }
+}
+
+MyResource r = new MyResource("bucket");

--- a/test/translations/classes/constructor_parameter_shadows_field.py
+++ b/test/translations/classes/constructor_parameter_shadows_field.py
@@ -1,0 +1,10 @@
+class MyResource:
+
+    def __init__(self, resource):
+        self.resource = resource
+        self.env = "production"
+
+    def description(self):
+        return self.resource
+
+r = MyResource("bucket")

--- a/test/translations/classes/constructor_parameter_shadows_field.ts
+++ b/test/translations/classes/constructor_parameter_shadows_field.ts
@@ -1,0 +1,15 @@
+class MyResource {
+  public readonly env: string;
+  private readonly resource: string;
+
+  constructor(resource: string) {
+    this.resource = resource;
+    this.env = 'production';
+  }
+
+  public description(): string {
+    return this.resource;
+  }
+}
+
+const r = new MyResource('bucket');

--- a/test/translations/expressions/assignment_to_parameter.cs
+++ b/test/translations/expressions/assignment_to_parameter.cs
@@ -1,0 +1,13 @@
+public string Normalize(string value, string fallback, int retries)
+{
+    if (!value)
+    {
+        value = fallback;
+    }
+    if (!fallback)
+    {
+        value = "anonymous";
+        retries = 3;
+    }
+    return value;
+}

--- a/test/translations/expressions/assignment_to_parameter.go
+++ b/test/translations/expressions/assignment_to_parameter.go
@@ -1,0 +1,10 @@
+func normalize(value *string, fallback *string, retries *f64) *string {
+	if !*value {
+		*value = *fallback
+	}
+	if !*fallback {
+		*value = "anonymous"
+		*retries = 3
+	}
+	return *value
+}

--- a/test/translations/expressions/assignment_to_parameter.java
+++ b/test/translations/expressions/assignment_to_parameter.java
@@ -1,0 +1,10 @@
+public String normalize(String value, String fallback, Number retries) {
+    if (!value) {
+        value = fallback;
+    }
+    if (!fallback) {
+        value = "anonymous";
+        retries = 3;
+    }
+    return value;
+}

--- a/test/translations/expressions/assignment_to_parameter.py
+++ b/test/translations/expressions/assignment_to_parameter.py
@@ -1,0 +1,7 @@
+def normalize(value, fallback, retries):
+    if not value:
+        value = fallback
+    if not fallback:
+        value = "anonymous"
+        retries = 3
+    return value

--- a/test/translations/expressions/assignment_to_parameter.ts
+++ b/test/translations/expressions/assignment_to_parameter.ts
@@ -1,0 +1,10 @@
+function normalize(value: string, fallback: string, retries: number): string {
+  if (!value) {
+    value = fallback;
+  }
+  if (!fallback) {
+    value = 'anonymous';
+    retries = 3;
+  }
+  return value;
+}

